rezervo/providers/sats/provider.py

```diff
@@ -0,0 +1,311 @@
+    SatsBookingsResponse,
+    SatsClass,
+    SatsDaySchedule,
+    SatsLocationIdentifier,
+    SatsScheduleResponse,
+    SatsWeekSchedule,
+)
+from rezervo.providers.sats.urls import (
+    BOOKING_URL,
+    BOOKINGS_PATH,
+    BOOKINGS_URL,
+    UNBOOK_URL,
+    schedule_url,
+)
+from rezervo.providers.schema import LocationIdentifier
+from rezervo.schemas.config.user import (
+    ChainUser,
+    ChainUserCredentials,
+    Class,
+    ClassTime,
+)
+from rezervo.schemas.schedule import (
+    RezervoActivity,
+    RezervoClass,
+    RezervoDay,
+    RezervoInstructor,
+    RezervoLocation,
+    RezervoSchedule,
+    SessionRezervoClass,
+    UserSession,
+)
+from rezervo.utils.category_utils import determine_activity_category
+
+
+class SatsProvider(Provider[ClientSession, SatsLocationIdentifier]):
+    async def _authenticate(
+        self, chain_user: ChainUser
+    ) -> Union[ClientSession, AuthenticationError]:
+        session = create_client_session()
+        return await authenticate_session(
+            session, chain_user.username, chain_user.password
+        )
+
+    async def find_class_by_id(
+        self, class_id: str
+    ) -> Union[RezervoClass, BookingError, AuthenticationError]:
+        provider_location_id_match = re.search(r"(\d+)p", class_id)
+        if not provider_location_id_match:
+            raise BookingError.MALFORMED_CLASS
+        location_id = self.location_from_provider_location_identifier(
+            int(provider_location_id_match.group(1))
+        )
+        if location_id is None:
+            raise BookingError.MALFORMED_CLASS
+        schedule = await self.fetch_schedule(datetime.datetime.now(), 14, [location_id])
+        for day in schedule.days:
+            for _class in day.classes:
+                if _class.id == class_id:
+                    return _class
+        return BookingError.CLASS_MISSING
+
+    async def find_class(
+        self, _class_config: Class
+    ) -> Union[RezervoClass, BookingError, AuthenticationError]:
+        schedule = await self.fetch_schedule(
+            datetime.datetime.now(), 14, [_class_config.location_id]
+        )
+        for day in schedule.days:
+            for _class in day.classes:
+                if (
+                    _class.activity.id == _class_config.activity_id
+                    and _class.start_time.weekday() == _class_config.weekday
+                    and _class.location.id == _class_config.location_id
+                    and _class.start_time.hour == _class_config.start_time.hour
+                    and _class.start_time.minute == _class_config.start_time.minute
+                ):
+                    return _class
+        raise BookingError.CLASS_MISSING
+
+    async def _book_class(
+        self,
+        auth_session: ClientSession,
+        class_id: str,
+    ) -> bool:
+        res = await auth_session.post(BOOKING_URL, data={"id": class_id})
+        await auth_session.close()
```

Comment:
Why `close` the session here?

Reply:
Got a warning when I did not close it, so I thought it was dandy to just do it manually. 🤠